In [ ]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

1. Lance une requête sur la page web indiquée ci-dessus.

2. Quel est le response code ? A quoi correspond cette erreur ? Chuck t'a grillé !

In [ ]:
url = 'https://chucknorrisfacts.net/facts.php?page=1'
page = requests.get(url)
print(page)

<Response [406]>


406 = NOT ACCEPTABLE

3. Contourne la protection en indiquant un navigateur existant. Et relance ta requête.

In [ ]:
nav = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.93 Safari/537.36"

page = requests.get(url, headers={'User-Agent': nav})
print(page)

<Response [200]>


4. Trouve la balise permettant d'identifier le bloc contenant une blague. Combien y en a-t-il ? (il devrait y en avoir 20). Affiche le 8e bloc de blague grâce à l'attribut .text

In [ ]:
# object beautifulSoup
soup = BeautifulSoup(page.content, 'html.parser')

# find the right 'balise'
contenu = soup.find(id='content')

# the jokes don't have 'class or id balise' : let's find the <p> in the <div id = 'content'>
jokes = contenu.find_all('p')

# how many jokes ?
len(jokes)

20

On pouvait aussi partir de la balise avec le style CSS mais le .text ne rendait pas aussi bien :

soup.find_all('div', {"style" : "border-top: 1px solid #251836; padding: 0 0 5px 7px;"})

In [ ]:
# 8th joke :
jokes[7].get_text()

'Jesus can walk on water, but Chuck Norris can swim through dry land.'

6. Fais de même pour isoler uniquement la note de la 8e blague

In [ ]:
notes = contenu.find_all(class_ ='out5Class')
notes[7].get_text()

'4.04'

7. Crée un dictionnaire vide, puis crée une boucle qui viendra enregistrer dans ce dictionnaire chaque blague en clé et chaque note correspondante en valeur.

In [ ]:
jokes_list=[]
for joke in jokes :
    jokes_list.append(joke.get_text())

notes_list= []
for note in notes:
    notes_list.append(note.get_text())

jokes_notes_dict = dict(zip(jokes_list, notes_list))
jokes_notes_dict

{'Chuck Norris can unscramble an egg.': '4.06',
 'Chuck Norris destroyed the periodic table, because he only recognizes the element of surprise.': '3.93',
 "Chuck Norris doesn't read books. He stares them down until he gets the information he wants.": '3.83',
 'Chuck Norris doesnt sleep. He waits.': '3.82',
 'Chuck Norris has a vacation home on the sun.': '3.21',
 'Chuck Norris is suing Myspace for taking the name of what he calls everything around you.': '3.38',
 'Chuck Norris is the only person in the world that can actually email a roundhouse kick.': '3.57',
 'Chuck Norris once shot an enemy plane down with his finger by yelling, "Bang!"': '3.94',
 "Chuck Norris' calendar goes straight from March 31st to April 2nd. No one fools Chuck Norris. ": '4.08',
 "Chuck Norris' tears cure cancer. Too bad he has never cried.": '4.04',
 'If you make a list of 10 things Chuck Norris cannot do, he will appear at your house and perform them all. Your life may be forfeit.': '3.53',
 'If you spell C

8. Transforme ce dictionnaire en DataFrame de 2 colonnes : blague et note. Il doit avoir 20 lignes : une par blague.

In [ ]:
df = pd.DataFrame({
    'Blagues' : jokes_notes_dict.keys(),
    'Note' : jokes_notes_dict.values()
})

df

Blagues  Note
0   There is no 'ctrl' button on Chuck Norris' com...  3.74
1                Chuck Norris doesnt sleep. He waits.  3.82
2   Some kids piss their name in the snow. Chuck N...  4.05
3   Chuck Norris' calendar goes straight from Marc...  4.08
4   Chuck Norris is the only person in the world t...  3.57
5   Chuck Norris' tears cure cancer. Too bad he ha...  4.04
6   When Chuck Norris is put in a straight jacket ...  3.37
7   Jesus can walk on water, but Chuck Norris can ...  4.04
8   Chuck Norris once shot an enemy plane down wit...  3.94
9   Chuck Norris doesn't read books. He stares the...  3.83
10  The quickest way to a man's heart is with Chuc...  3.82
11  If you spell Chuck Norris in Scrabble, you win...  3.96
12  The following is a short list of things Chuck ...  3.89
13  If you make a list of 10 things Chuck Norris c...  3.53
14  When Chuck Norris looks in the mirror nothing ...  3.92
15                Chuck Norris can unscramble an egg.  4.06
16  When Chuck Norris wants popcorn, he breathes o...  3.33
17       Chuck Norris has a vacation home on the sun.  3.21
18  Chuck Norris is suing Myspace for taking the n...  3.38
19  Chuck Norris destroyed the periodic table, bec...  3.93